src/parser/ion_1_0/text/mod.rs

```diff
@@ -0,0 +1,2614 @@
+
+enum FloatOrDecimal {
+    Float(String),
+    Decimal(String),
+}
+
+fn take_float_or_decimal(i: &str) -> IonResult<&str, ion::Data> {
+    let (i, integer) = take_dec_integer(i)?;
+    let (i, fractional) = opt(take_dec_frac)(i)?;
+    let (i, exponent) = opt(alt((
+        map(take_float_exp, FloatOrDecimal::Float),
+        map(take_decimal_exp, FloatOrDecimal::Decimal),
+    )))(i)?;
+
+    // If there is no fractional and no exponent part then this is an integer.
+    if fractional.is_none() && exponent.is_none() {
+        return Err(Err::Error(IonError::from_error_kind(i, ErrorKind::Digit)));
+    }
+
+    match exponent {
+        Some(FloatOrDecimal::Float(exponent)) => assemble_float(i, integer, fractional, exponent),
+        Some(FloatOrDecimal::Decimal(exponent)) => {
+            assemble_decimal(i, integer, fractional, Some(exponent))
+        }
+        None => assemble_decimal(i, integer, fractional, None),
+    }
+}
+
+/// Note: number parsing consolidated to avoid parsing DEC_INTEGER multiple times when deciding
+/// the numeric type to apply.
+///
+/// FLOAT
+///     : DEC_INTEGER DEC_FRAC? FLOAT_EXP
+///     ;
+fn assemble_float<'a>(
+    i: &'a str,
+    integer: String,
+    fractional: Option<Vec<&'a str>>,
+    exponent: String,
+) -> IonResult<&'a str, ion::Data> {
+    let mut float = integer;
+
+    if let Some(fractional) = fractional {
+        float.push('.');
+        fractional.iter().for_each(|s| float.push_str(s));
+    }
+
+    float.push('e');
+    float.push_str(&exponent);
+
+    match float.parse::<f64>() {
+        Ok(f) => Ok((i, ion::Data::Float(Some(f)))),
+        Err(e) => Err(Err::Failure(IonError::from_format_error(
+            i,
+            FormatError::Text(TextFormatError::FloatParse(float, e)),
+        ))),
+    }
+}
+
+/// fragment
+/// FLOAT_EXP
+///     : [Ee] PLUS_OR_MINUS? DEC_DIGIT+
+///     ;
+fn take_float_exp(i: &str) -> IonResult<&str, String> {
+    take_exp("eE")(i)
+}
+
+fn take_exp(exponent_delimiters: &'static str) -> impl Fn(&str) -> IonResult<&str, String> {
+    move |i: &str| {
+        let (i, _) = one_of(exponent_delimiters)(i)?;
+        let (i, sign) = opt(take_plus_or_minus)(i)?;
+        let (i, digits) = take_while1(|c: char| c.is_ascii_digit())(i)?;
+
+        let sign_char = if sign.is_some() { 1 } else { 0 };
+
+        let mut exponent = String::with_capacity(digits.len() + sign_char);
+        if let Some(sign) = sign {
+            exponent.push(sign);
+        }
+        exponent.push_str(digits);
+
+        Ok((i, exponent))
+    }
+}
+
+///
+/// Encoding Section: Ion Decimal
+///
+
+/// Note: number parsing consolidated to avoid parsing DEC_INTEGER multiple times when deciding
+/// the numeric type to apply.
+///
+/// DECIMAL
+///     : DEC_INTEGER DEC_FRAC? DECIMAL_EXP?
+///     ;
+fn assemble_decimal<'a>(
+    i: &'a str,
+    integer: String,
+    fractional: Option<Vec<&'a str>>,
+    exponent: Option<String>,
+) -> IonResult<&'a str, ion::Data> {
```

Comment:
This could more specifically return `IonResult<&'a str, ion::Decimal>`

Reply:
Adjusted this one and `take_float` to return an `f64`.

On a related note, how do you feel about a blanket Into<Data> impl for value types?

Reply:
That seems reasonable to me, though I think by convention `From` is preferred?

Reply:
Yep, just harder to enumerate that way. :)